group_by.py

```diff
@@ -0,0 +1,19 @@
+def group_by(function, iterable):
+    """
+    :param function: A function to execute on each iterable element.
+    :param iterable: An iterable object.
+    :return: A dictionary that its keys is the results of the function on the elements and the values is the elements.
+    """
+    function_results_dict = {}
+
+    for element in iterable:
+        res = function(element)
+        if function_results_dict.get(res):
+            function_results_dict[res] += [element]
+        else:
+            function_results_dict[res] = [element]
```

Comment:
I would recommend using default_dict - an elegant and efficient way of inserting unknown keys into a dictionary.
https://realpython.com/python-defaultdict/#:~:text=However%2C%20the%20defaultdict%20version%20is,instead%20of%20a%20standard%20dict%20.